velox/common/base/AsyncSource.h

```diff
@@ -144,6 +144,39 @@ class AsyncSource {
     return timing_;
   }
 
+  /// This function assists the caller in ensuring that resources allocated in
+  /// AsyncSource are promptly released:
+  /// 1. Waits for the completion of the 'make_' function if it is executing in
+  /// the thread pool.
+  /// 2. Resets the 'make_' function if it has not started yet.
+  /// 3. Cleans up the 'item_' if 'make_' has completed, but the result 'item_'
+  /// has not been returned to the caller.
+  void close() {
+    if (closed_) {
+      return;
+    }
+    ContinueFuture wait;
+    {
+      std::lock_guard<std::mutex> l(mutex_);
+      if (making_ && !promise_) {
+        promise_ = std::make_unique<ContinuePromise>();
+        wait = promise_->getSemiFuture();
+      }
+      if (make_) {
+        make_ = nullptr;
+      }
+    }
+
+    auto& exec = folly::QueuedImmediateExecutor::instance();
+    std::move(wait).via(&exec).wait();
```

Comment:
A better way is to cancel the pending request then wait in case it's loading a large block of useless data, but currently it's not possible on S3 since it uses sync GetObject

Reply:
Agree, this is the best way, but we need all dependencies to support cancellation.
From my understanding, waiting only happens when the task is failed or cancelled.

Reply:
Would it make sense to introduce a "cancel" API and have some implementations "wait" while others actually stop processing?

Reply:
Perhaps we could enhance the current solution by introducing a parameter to signal to the `make_` function that it should abort. Subsequently, each implementation of `make_` can appropriately respond to this signal. This concept mirrors the functionality of [CancellationToken](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-8.0) in .NET. However, I'm not sure if C++ or folly offer similar features.

new proposed signature looks like: `std::function<std::unique_ptr<Item>(CancellationToken)> make_;`

Reply:
We can add this later I think, in the change we mainly solve the memory leak.